modifying docker logs --since command to include nanoseconds

mikebrow opened the pull request:
Added ".nanoseconds" to since variable in json that is being passed between client and server for `docker log -t --since` command. 

Fix #16396
Fix #16395 

Signed-off-by: Mike Brown brownwm@us.ibm.com

mikebrow committed:
modifying docker logs --since command to include nanoseconds
and Adding support for fractions of a second of any length;
Also fix timezone issues and update unit test cases

Signed-off-by: Mike Brown <brownwm@us.ibm.com>

mikebrow committed:
modifying docker logs --since command to include nanoseconds
and Adding support for fractions of a second of any length;
Also fix timezone issues and update unit test cases

Signed-off-by: Mike Brown <brownwm@us.ibm.com>

phemmer commented:
I don't think this is what you meant to do.
This is going to treat `123.456` as 123 seconds and 456 nanoseconds. When really `123.456` is 123 seconds and 456 milliseconds (456000000 nanoseconds).

mikebrow commented:
I see your point, but the format of putting the `.` between the seconds and nanoseconds is only for convenience of using one string as a parameter and the `.` as the separator.... it's not meant to indicate that it's 456 milliseconds. On both sides it's 456 nanoseconds.  Put another way .. it's since = `seconds.nanoseconds`

phemmer commented:
I recognize that as long as both sides treat it the same, it'll behave properly, but in my opinion, the behavior is not intuitive.
If I were writing a client to communicate with docker via the API, I would expect a value of `123.456` to mean 123 seconds and 456 milliseconds after unix epoch.
If 456 is meant to be nanoseconds, it should be provided as a separate parameter.

mikebrow commented:
Options:
1) prefix leading zeros for the nanoseconds so it's more intuitive.
2) add a new parameter (client and server) for the json being passed sincenanoseconds? Rename since to sinceseconds? 
3) change the separator I used from `.` to `,` 
...
I suggest 1 above...

duglin commented:
I long time ago I remember doing some work with this kind of stuff and I dealing with the nano part was a bit of a pain. I tried to google the docs I used to help clarify this for me but had not luck, but my vague recollection was that leading (or was it trailing?) zeros were important.  If 1 is what the official rfc/docs say, then I vote for 1 as well.  btw - do you have a url to the spec for this stuff?

mikebrow commented:
There is no spec for the format of the since parameter.  But it makes sense to prefix with leading zeroes and to account for the possibility that someone might use a different fraction type when passing since between the client and server... So I'll go ahead and use the for mat `%d.%09d`,seconds,nanoseconds for the client and parse whatever is receive appropriately from the server.  Meaning: will support milliseconds, nanoseconds, picoseconds... and correct to nanoseconds when creating the timestamp per the time variable in go.

mikebrow commented:
@phemmer @duglin  I've modified the PR to include the client outputting with the suggested format and the server receiving a fraction of a second to any precision, milliseconds, nanoseconds, picoseconds etc.

mikebrow commented:
Based on comments from @duglin I've gone ahead and fixed other identified issues with the code.  Namely invalid formats for timestamps were not failing in the client, so the error message was from the server not the client and the error message from the server did not explain the problem.  Additionally, when using RFC3339 and RFC3339Nano timestamps we were incorrectly processing timestamps that did not have one of `Z` or `+00:00` formats.  RFC3339 and *Nano requires that the timezone be specified.  We were doing this but truncating the RFC3339 format and assuming local time zone on the time stamps.  However the server may not be in the local time zone and the logs report in UTC ... thus the behavior was not well defined.

mikebrow commented:
Also fixes issue #16395

runcom commented:
You should add that to the first post like Fix #PRNUMBER in order to auto close them when this gets merged

phemmer commented on api/server/router/local/container.go:
```diff
@@ -104,11 +105,29 @@ func (s *router) getContainersLogs(ctx context.Context, w http.ResponseWriter, r
+			if err != nil {
+				return err
+			}
+			since = time.Unix(s, 0)
```
Can dedup this a little by doing something like:

```
sa := strings.SplitN(r.Form.Get("since"), 2)
s, err := strconv.ParseInt(sa[0], 10, 64)
if err != nil {
 return err
}
var n int64
if len(sa) == 2 {
 n, err := strconv.ParseInt(sa[1], 10, 64)
 if err != nil {
  return err
 }
}
since = time.Unix(s, n)
```

mikebrow commented:
Per discussion with @duglin  I've refactored the code, added a new function for parsing the json received unix timestamp, added testcases for the new function, added additional supported formats for the command line, and update the docs to reflect the changes.

phemmer commented on pkg/timeutils/utils.go:
```diff
@@ -1,36 +1,102 @@
+		return s, 0, nil
 	}
-	return strconv.FormatInt(t.Unix(), 10)
+	return def, 0, nil
```
just a minor nit, you can avoid a layer of indentation by doing something like:

```
if value == "" {
 return def, 0, nil
}
sa := strings.SplitN(value, ".", 2)
...
```

https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

mikebrow replied:
Nits addressed... cheers.

duglin commented on api/server/router/local/container.go:
```diff
@@ -17,6 +17,7 @@ import (
 	derr "github.com/docker/docker/errors"
 	"github.com/docker/docker/pkg/ioutils"
 	"github.com/docker/docker/pkg/signal"
+	tu "github.com/docker/docker/pkg/timeutils"
```
unless there's a conflict we should probably just drop the alias.

mikebrow replied:
alias dropped and replaced with packagename. references

mikebrow commented:
Per discussion with @duglin I've added additional formats for partial time entries... so now we will support T00:00:00, T00:00, and T00 with and without a specified timezone Z00:00 on the end.   And added enough testcases to get to 100% coverage on the two apis in the timeutils package.

duglin commented on pkg/timeutils/utils_test.go:
```diff
@@ -36,9 +52,37 @@ func TestGetTimestamp(t *testing.T) {
+		o, err := GetTimestamp(c.in, now)
 		if o != c.expected {
-			t.Fatalf("wrong value for '%s'. expected:'%s' got:'%s'", c.in, c.expected, o)
+			if err != nil {
```
Are there any error tests we want to include? perhaps the tests you removed - like "0" ?

duglin replied:
never mind

duglin commented on pkg/timeutils/utils_test.go:
```diff
@@ -9,22 +9,38 @@ import (
 func TestGetTimestamp(t *testing.T) {
 	now := time.Now()
 	cases := []struct{ in, expected string }{
-		{"0", "-62167305600"}, // 0 gets parsed year 0
```
Reviewers - make sure you're ok with dropping support for some of these formats. Some are pretty weird.

duglin commented on pkg/timeutils/utils.go:
```diff
@@ -1,36 +1,122 @@
 	}
-	return strconv.FormatInt(t.Unix(), 10)
+	// should already be in nanoseconds but just in case convert n to nanoseonds
+	n *= int64(math.Pow(float64(10), float64(9-len(sa[1]))))
```
not sure what happens when len(sa[1]) is greater than 9 - seems like we should just substr in those cases instead of letting it do 10^-20

mikebrow replied:
Note the unit testcases that were added:
		{"1136073600.000000001", 0, 1136073600, 1},
		{"1136073600.0000000010", 0, 1136073600, 1},
		{"1136073600.00000001", 0, 1136073600, 10},

The target output is in nanoseconds because that's the variable needed for the unixnano time function, if the length is > 9 the code above shifts the decimal place to the left, if the length is less than 9 it shifts the decimal place to the right (mathematically).  Could've done this with string functions instead of math by truncating (substr) it to 9 digits if len(sa[1])>9 and adding the right number of zeros to the end to make it 9 in length if len(sa[1])< 9 before converting sa[1] to an int64...

duglin commented:
just one minor comment.
ping @cpuguy83 @vdemeester @runcom @jfrazelle

hqhq committed:
Bump bolt to v1.1.0

It adds ARM64, ppc64le, s390x, solaris support, and a bunch of
bugfixs.

Signed-off-by: Qiang Huang <h.huangqiang@huawei.com>

coolljt0725 committed:
Fix docker inspect display odd gateway value for none network mode

Signed-off-by: Lei Jitang <leijitang@huawei.com>

LK4D4 committed:
Do not ignore error from driver.Put

Signed-off-by: Alexander Morozov <lk4d4@docker.com>

thaJeztah commented:
ping @cpuguy83 @vdemeester ptal

duglin commented:
it was implied, but to be explicit  ;-)   LGTM

calavera committed:
Merge pull request #17445 from tiborvass/fix-docker-py

docker-py: upgrade and fix test script

calavera committed:
Merge pull request #17424 from coolljt0725/fix_inspect_gateway_value

Fix docker inspect display odd gateway value for none network mode

vdemeester commented on pkg/timeutils/utils_test.go:
```diff
@@ -36,9 +52,37 @@ func TestGetTimestamp(t *testing.T) {
+		{"1136073600.000000001", 0, 1136073600, 1},
+		{"1136073600.0000000010", 0, 1136073600, 1},
+		{"1136073600.00000001", 0, 1136073600, 10},
+		{"foo.bar", 0, 0, 0},
```
Not sure I understand this test but, shouldn't `ParseTimestamps` return `err` in this case ?

mikebrow replied:
It does return an err but the test case is not comparing expected error values its only comparing the expected results.

vdemeester replied:
oh, ok :)
Then I find the test below a little strange : if the `expectedS` or `expectedN` are different but there is not an error returned, the test (or at least this case) passes ?

mikebrow replied:
Correct.  IOW I need to fix that ... it should look like line 60 above.

LK4D4 committed:
Remove unnecessary error returns

Signed-off-by: Alexander Morozov <lk4d4@docker.com>

LK4D4 committed:
Fix docker-device-tool

Signed-off-by: Alexander Morozov <lk4d4@docker.com>

mikebrow committed:
Merge branch 'docker-logs-since-ignores-nanoseconds' of https://github.com/mikebrow/docker into docker-logs-since-ignores-nanoseconds

Signed-off-by: Mike Brown <brownwm@us.ibm.com>

Conflicts:
	docs/reference/commandline/events.md
	pkg/timeutils/utils.go
	pkg/timeutils/utils_test.go

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "docker-logs-since-ignores-nanoseconds" git@github.com:mikebrow/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~10
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

mikebrow commented:
Will reopen this PR with a clean commit.

phemmer commented:
Just fyi, you can do `git push --force` instead of opening a new PR.

mikebrow commented:
@phemmer Thx will try that next time.
